openassessment/assessment/models/staff.py

```diff
@@ -0,0 +1,117 @@
+    course_id = models.CharField(max_length=40, db_index=True)
+    item_id = models.CharField(max_length=128, db_index=True)
+    submission_uuid = models.CharField(max_length=128, db_index=True, unique=True)
+    created_at = models.DateTimeField(default=now, db_index=True)
+    grading_completed_at = models.DateTimeField(null=True, db_index=True)
+    grading_started_at = models.DateTimeField(null=True, db_index=True)
+    cancelled_at = models.DateTimeField(null=True, db_index=True)
+    assessment = models.CharField(max_length=128, db_index=True, null=True)
+
+    class Meta:
+        ordering = ["created_at", "id"]
+        app_label = "assessment"
+
+    @property
+    def is_cancelled(self):
+        """
+        Check if the workflow is cancelled.
+
+        Returns:
+            True/False
+        """
+        return bool(self.cancelled_at)
+
+    @classmethod
+    def get_submission_for_review(cls, course_id, item_id, scorer_id):
+        """
+        Find a submission for peer assessment. This function will find the next
+        submission that requires assessment, excluding any submission that has been
+        completely graded, or is actively being reviewed by other students.
+
+        Args:
+            submission_uuid (str): The submission UUID from the student
+                requesting a submission for assessment. This is used to explicitly
+                avoid giving the student their own submission, and determines the
+                associated Peer Workflow.
+            item_id (str): The student_item that we would like to retrieve submissions for.
+            scorer_id (str): The user id of the staff member scoring this submission
+
+        Returns:
+            submission_uuid (str): The submission_uuid for the submission to review.
+
+        Raises:
+            StaffAssessmentInternalError: Raised when there is an error retrieving
+                the workflows for this request.
+
+        """
+        timeout = (now() - cls.TIME_LIMIT).strftime("%Y-%m-%d %H:%M:%S")
+        try:
+            # Search for existing submissions that the scorer has worked on.
+            staff_workflows = StaffWorkflow.objects.filter(
+                course_id=course_id,
+                item_id=item_id,
+                scorer_id=scorer_id,
+                grading_completed_at=None,
+                cancelled_at=None,
+            )
+            # If no existing submissions exist, then get any other
+            # available workflows.
+            if not staff_workflows:
+                staff_workflows = StaffWorkflow.objects.filter(
+                    models.Q(scorer_id='') | models.Q(grading_started_at__lte=timeout),
+                    course_id=course_id,
+                    item_id=item_id,
+                    grading_completed_at=None,
+                    cancelled_at=None,
+                )
+            if not staff_workflows:
+                return None
+
+            workflow = staff_workflows[0]
```

Comment:
If there is a race condition here where multiple another staff might overwrite after a first staff initiates grading, do we protect against this later?  Or can the original staff possibly update with the wrong scorer_id?